// react-title-component/es6/Title.js
'use strict';

export { flushTitle };
import React from 'react';

var titles = [];

function getTitle() {
  return titles[titles.length - 1];
}

function updateTitle() {
  document.title = getTitle();
}

function flushTitle() {
  var title = getTitle();
  titles = [];
  return title;
}

var _React$PropTypes = React.PropTypes;
var oneOfType = _React$PropTypes.oneOfType;
var string = _React$PropTypes.string;
var func = _React$PropTypes.func;

var Title = React.createClass({
  displayName: 'Title',

  propTypes: {
    render: oneOfType([string, func]).isRequired
  },

  getInitialState: function getInitialState() {
    return {
      index: titles.push('') - 1
    };
  },

  componentWillUnmount: function componentWillUnmount() {
    titles.pop();
  },

  componentDidMount: updateTitle,

  componentDidUpdate: updateTitle,

  render: function render() {
    var render = this.props.render;

    titles[this.state.index] = typeof render === 'function' ? render(titles[this.state.index - 1] || '') : render;
    return this.props.children || null;
  }

});

export default Title;
